Update gefs headline plot scripts

BinbinZhou-NOAA opened the pull request:
## Pull Request Testing ##

- [ ] Describe testing already performed for this Pull Request:</br>

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

- [ ] Has the code been checked to ensure that no errors occur during the execution? **[Yes or No]**

- [ ] Do these updates/additions include sufficient testing updates? **[Yes or No]**

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##

- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR description above.
- [ ] Ensure the PR title matches the feature branch name.
- [ ] Check the following:
- [ ]  Instructions provided on how to run
- [ ]  Developer's name is replaced by ${user} where necessary throughout the code
- [ ]  Check that the ecf file has all the proper definitions of variables
- [ ]  Check that the jobs file has all the proper settings of COMIN and COMOUT and other input variables
- [ ]  Check to see that the output directory structure is followed
- [ ]  Be sure that you are not using MET utilities outside the METplus wrapper structure

- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue.

BinbinZhou-NOAA committed:
Update gefs headline plot scripts

AliciaBentley-NOAA commented on dev/drivers/scripts/plots/global_ens/jevs_global_ens_headline_gefs_grid2grid_plots.sh:
```diff
@@ -7,6 +7,16 @@
 
+#************************************************************************
+#Note: (1) This headline plot generation driver script should be run 
+#           on Jan 18th of each year. 
```
@BinbinZhou-NOAA Please say that this should be run on Jan 16th of each year. Congress has asks for this earlier than the Jan 18th (as they did this year). We can manually get the forecast/analysis files earlier than EVS gets it if we need if we have to. 

CC @PerryShafran-NOAA @ShelleyMelchior-NOAA

BinbinZhou-NOAA committed:
Update comments in the headline plot driver script

BinbinZhou-NOAA commented:
Alicia,

  OK, the comment in the headline plot driver script has been updated.

Thanks!

Binbin

On Fri, Jan 19, 2024 at 12:07 PM Alicia Bentley ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> dev/drivers/scripts/plots/global_ens/jevs_global_ens_headline_gefs_grid2grid_plots.sh
> <https://github.com/NOAA-EMC/EVS/pull/415#discussion_r1459362163>:
>
> > @@ -7,6 +7,16 @@
>  #PBS -l place=shared,select=1:ncpus=1:mem=5GB
>  #PBS -l debug=true
>
> +#************************************************************************
> +#Note: (1) This headline plot generation driver script should be run
> +#           on Jan 18th of each year.
>
> @BinbinZhou-NOAA <https://github.com/BinbinZhou-NOAA> Please say that
> this should be run on Jan 16th of each year. Congress has asks for this
> earlier than the Jan 18th (as they did this year). We can manually get the
> forecast/analysis files earlier than EVS gets it if we need if we have to.
>
> CC @PerryShafran-NOAA <https://github.com/PerryShafran-NOAA>
> @ShelleyMelchior-NOAA <https://github.com/ShelleyMelchior-NOAA>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EVS/pull/415#pullrequestreview-1833276952>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQBMPFMZ2AUI5CM2CV5EX7DYPKR3HAVCNFSM6AAAAABCCHRNHSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQMZTGI3TMOJVGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Binbin Zhou

Physical Scientist

Lynker at NOAA/NWS/NCEP/EMC

5830 University Research Ct.

College Park, MD 20740

***@***.***

301-683-3683

PerryShafran-NOAA commented:
@BinbinZhou-NOAA I have your clone, and because I have all the data I need on *Dogwood*, I am running the test there as emc.vpppg.  I will let you know when the job is done.

ShelleyMelchior-NOAA commented on scripts/plots/global_ens/exevs_global_ens_headline_gefs_grid2grid_plots.sh:
```diff
@@ -3,24 +3,25 @@
-#  Note: 1. The headline score needs the stat files over 0ne year + 17 days 
-#           (this year + next 17 days of Jan) 
-#        2. This job is run on Jan 17 of each year                 
+#  Note: 1. The headline score needs the stat files over 0ne year + 16 days 
```
@BinbinZhou-NOAA , can you correct that "0" (zero) to "O" (capital O) instead? Nitpicky, but would be a good correction.

BinbinZhou-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/headline_bugfix

BinbinZhou-NOAA committed:
Fix another running day in the comments and a typo

BinbinZhou-NOAA committed:
fix another running day in dirver script

BinbinZhou-NOAA committed:
Fix a typo in the headline plot script

BinbinZhou-NOAA commented:
Shelley,

  It was fixed,

Thanks!

Binbin

On Fri, Jan 19, 2024 at 2:01 PM Shelley Melchior ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> scripts/plots/global_ens/exevs_global_ens_headline_gefs_grid2grid_plots.sh
> <https://github.com/NOAA-EMC/EVS/pull/415#discussion_r1459569064>:
>
> > @@ -3,24 +3,25 @@
>  # Purpose:  1. Setup the environment parameter for running GEFS headline plotting script
>  #           2. Run plotting script $USHevs/global_ens/evs_global_ens_headline_plot.sh
>  #
> -#  Note: 1. The headline score needs the stat files over 0ne year + 17 days
> -#           (this year + next 17 days of Jan)
> -#        2. This job is run on Jan 17 of each year
> +#  Note: 1. The headline score needs the stat files over 0ne year + 16 days
>
> @BinbinZhou-NOAA <https://github.com/BinbinZhou-NOAA> , can you correct
> that "0" (zero) to "O" (capital O) instead? Nitpicky, but would be a good
> correction.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EVS/pull/415#pullrequestreview-1833491162>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQBMPFJELSNIV7ANRCDMJRTYPK7GFAVCNFSM6AAAAABCCHRNHSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQMZTGQ4TCMJWGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Binbin Zhou

Physical Scientist

Lynker at NOAA/NWS/NCEP/EMC

5830 University Research Ct.

College Park, MD 20740

***@***.***

301-683-3683

PerryShafran-NOAA commented on ush/global_ens/evs_global_ens_headline_plot.sh:
```diff
@@ -126,7 +153,7 @@ for model in gfs gefs naefs ; do
-   # stat_analysis -lookin $stat_data/$model/${model}_headline_*${yyyy}*.stat -fcst_valid_hour 00  -job aggregate_stat -line_type SAL1L2 -out_line_type CNT -by FCST_VAR,FCST_LEV,FCST_LEAD,VX_MASK -out_stat agg_stat_SAL1L2_to_CNT.${model}.${yyyy}.00Z
+    stat_analysis -lookin $stat_data/$model/${model}_headline_v*.stat -fcst_valid_hour 00  -job aggregate_stat -line_type SAL1L2 -out_line_type CNT -by FCST_VAR,FCST_LEV,FCST_LEAD,VX_MASK -out_stat agg_stat_SAL1L2_to_CNT.${model}.${yyyy}.00Z
 
    ${METPLUS_PATH}/ush/run_metplus.py -c ${PARMevs}/metplus_config/machine.conf -c ${PLOT_CONF}/StatAnlysis_fcstGENS_obsAnalysis_GatherByYear.conf 
```
@BinbinZhou-NOAA I think that this is causing an error in the execution of the code.  See the .o file: /lfs/h2/emc/vpppg/noscrub/emc.vpppg/pr415test/EVS/dev/drivers/scripts/plots/global_ens/jevs_global_ens_headline_gefs_grid2grid_plots.o125961829

```
WARNING: 
WARNING: get_filenames() -> can't stat "/lfs/h2/emc/stmp/emc.vpppg/evs_test/prod/tmp/jevs_global_ens_headline_gefs_grid2grid_plots.125961829
.dbqs01/all_stats/gfs/gfs_headline_v*.stat"
WARNING: 
ERROR  : 
ERROR  : process_search_dirs() -> no STAT files found in the directories specified!
ERROR  : 
ERROR  : 
ERROR  : main() -> encountered an error value of 1.  Calling clean_up() and usage() before exiting.
ERROR  :
```

The stat_analysis command should be looking for `${model}_headline_grid2grid_v*.stat`.  You're missing the grid2grid part of the file name.

PerryShafran-NOAA replied:
I am testing this with this code being `${model}_headline_${VERIF_CASE}_v*.stat`.

BinbinZhou-NOAA committed:
fix input stat file name string

BinbinZhou-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/headline_bugfix

BinbinZhou-NOAA commented:
Perry,

  The  stat_analysis -lookin $stat_data/$model/${model}_headline_v*.stat
should be
 stat_analysis -lookin $stat_data/$model/${model}_headline_grid2grid_v*.stat

I have fixed that and updated in github fork.

Thanks!

Binbin

On Fri, Jan 19, 2024 at 3:33 PM PerryShafran-NOAA ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In ush/global_ens/evs_global_ens_headline_plot.sh
> <https://github.com/NOAA-EMC/EVS/pull/415#discussion_r1459721293>:
>
> > @@ -126,7 +153,7 @@ for model in gfs gefs naefs ; do
>      export output_base=$DATA/plot
>      export stat_file_dir=$stat_data/$model
>
> -   # stat_analysis -lookin $stat_data/$model/${model}_headline_*${yyyy}*.stat -fcst_valid_hour 00  -job aggregate_stat -line_type SAL1L2 -out_line_type CNT -by FCST_VAR,FCST_LEV,FCST_LEAD,VX_MASK -out_stat agg_stat_SAL1L2_to_CNT.${model}.${yyyy}.00Z
> +    stat_analysis -lookin $stat_data/$model/${model}_headline_v*.stat -fcst_valid_hour 00  -job aggregate_stat -line_type SAL1L2 -out_line_type CNT -by FCST_VAR,FCST_LEV,FCST_LEAD,VX_MASK -out_stat agg_stat_SAL1L2_to_CNT.${model}.${yyyy}.00Z
>
>     ${METPLUS_PATH}/ush/run_metplus.py -c ${PARMevs}/metplus_config/machine.conf -c ${PLOT_CONF}/StatAnlysis_fcstGENS_obsAnalysis_GatherByYear.conf
>
> I am testing this with this code being
> ${model}_${VERIF_CASE}_headline.v*.stat.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EVS/pull/415#discussion_r1459721293>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQBMPFJRD443SZQ4ITJMEUTYPLKB5AVCNFSM6AAAAABCCHRNHSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQMZTGY2DAMJXGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Binbin Zhou

Physical Scientist

Lynker at NOAA/NWS/NCEP/EMC

5830 University Research Ct.

College Park, MD 20740

***@***.***

301-683-3683

AliciaBentley-NOAA commented:
@PerryShafran-NOAA  is going to work on this PR again tomorrow morning now that the dev machine is back- just FYI for @BinbinZhou-NOAA!

PerryShafran-NOAA commented:
@BinbinZhou-NOAA I am running now under emc.vpppg on Dogwood.  Will let you know the result.

PerryShafran-NOAA commented:
@BinbinZhou-NOAA I finished the global_ens headline plot job on Dogwood (current prod machine).  I set VDATE=20240116 (the new instructions seemed to indicate this is how to do it.)

Please look at the .o file: /lfs/h2/emc/vpppg/noscrub/emc.vpppg/pr415test/EVS/dev/drivers/scripts/plots/global_ens/jevs_global_ens_headline_gefs_grid2grid_plots.o127434803

And the output is here: /lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0/plots/global_ens/headline.20240116

I see there is a new text file, day_acc_below_0.6.txt.  Please check both the plot and the txt file to see that all is good.  

@AliciaBentley-NOAA @ShelleyMelchior-NOAA for awareness

PerryShafran-NOAA commented:
@BinbinZhou-NOAA I'm running this again because we noticed that the ACC <0.6 day was not exactly the same as last week.  I discovered that I might have set run_entire_yes to no in last week's run.  I'm checking to see if that made a difference.  Stay tuned for a sec.

I did change the directory where the output is located to headline.20240116_save.

BinbinZhou-NOAA commented:
Perry,

  The run_entire_yes must be set to yes, otherwise only dates in Jan are
covered.

Binbin


On Tue, Jan 23, 2024 at 10:57 AM PerryShafran-NOAA ***@***.***>
wrote:

> @BinbinZhou-NOAA <https://github.com/BinbinZhou-NOAA> I'm running this
> again because we noticed that the ACC <0.6 day was not exactly the same as
> last week. I discovered that I might have set run_entire_yes to no in last
> week's run. I'm checking to see if that made a difference. Stay tuned for a
> sec.
>
> I did change the directory where the output is located to
> headline.20240116_save.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EVS/pull/415#issuecomment-1906372632>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQBMPFIMZELXICQMBMD7OQ3YP7MX7AVCNFSM6AAAAABCCHRNHSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBWGM3TENRTGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Binbin Zhou

Physical Scientist

Lynker at NOAA/NWS/NCEP/EMC

5830 University Research Ct.

College Park, MD 20740

***@***.***

301-683-3683

PerryShafran-NOAA commented:
@BinbinZhou-NOAA I noticed that even if I set run_entire_year to NO, it's set to YES in the exevs script.  So it's always set to YES even if I set it to no in the driver, I think.

BinbinZhou-NOAA commented:
Perry,

  Show me your o file.

Binbin

On Tue, Jan 23, 2024 at 11:57 AM PerryShafran-NOAA ***@***.***>
wrote:

> @BinbinZhou-NOAA <https://github.com/BinbinZhou-NOAA> I noticed that even
> if I set run_entire_year to NO, it's set to YES in the exevs script. So
> it's always set to YES even if I set it to no in the driver, I think.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EVS/pull/415#issuecomment-1906510515>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQBMPFNRM3XWR55CXBFXK4TYP7TYXAVCNFSM6AAAAABCCHRNHSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBWGUYTANJRGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Binbin Zhou

Physical Scientist

Lynker at NOAA/NWS/NCEP/EMC

5830 University Research Ct.

College Park, MD 20740

***@***.***

301-683-3683

BinbinZhou-NOAA commented:
Perry,

  I found your o file that show the driver script on dogwood:
/lfs/h2/emc/vpppg/noscrub/emc.vpppg/pr415test/EVS/dev/drivers/scripts/plots/global_ens/jevs_global_ens_headline_gefs_grid2grid_plots.sh
In this script, export run_entire_year=no
Was this script that you ran?
If yes, please set it to yes.

Thanks!

Binbin


On Tue, Jan 23, 2024 at 1:27 PM Binbin Zhou - NOAA Affiliate <
***@***.***> wrote:

> Perry,
>
>   Show me your o file.
>
> Binbin
>
> On Tue, Jan 23, 2024 at 11:57 AM PerryShafran-NOAA <
> ***@***.***> wrote:
>
>> @BinbinZhou-NOAA <https://github.com/BinbinZhou-NOAA> I noticed that
>> even if I set run_entire_year to NO, it's set to YES in the exevs script.
>> So it's always set to YES even if I set it to no in the driver, I think.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/NOAA-EMC/EVS/pull/415#issuecomment-1906510515>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AQBMPFNRM3XWR55CXBFXK4TYP7TYXAVCNFSM6AAAAABCCHRNHSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBWGUYTANJRGU>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>
>
> --
>
> Binbin Zhou
>
> Physical Scientist
>
> Lynker at NOAA/NWS/NCEP/EMC
>
> 5830 University Research Ct.
>
> College Park, MD 20740
>
> ***@***.***
>
> 301-683-3683
>


-- 

Binbin Zhou

Physical Scientist

Lynker at NOAA/NWS/NCEP/EMC

5830 University Research Ct.

College Park, MD 20740

***@***.***

301-683-3683

PerryShafran-NOAA commented:
OK, I'm going to run that again ensuring that run_entire_year is set to yes in the driver script.  

I cannot get the ACC useful skill numbers from the run from last week, though.  Can you run this script and see what numbers you get?

BinbinZhou-NOAA commented:
Perry,

  OK, I'll run the driver script. Please show me the COMIN directory on
dogwood.

Thanks!

Binbin

On Tue, Jan 23, 2024 at 2:51 PM PerryShafran-NOAA ***@***.***>
wrote:

> OK, I'm going to run that again ensuring that run_entire_year is set to
> yes in the driver script.
>
> I cannot get the ACC useful skill numbers from the run from last week,
> though. Can you run this script and see what numbers you get?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EVS/pull/415#issuecomment-1906816241>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQBMPFJFDRKIBYR6RWU33YLYQAIEJAVCNFSM6AAAAABCCHRNHSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBWHAYTMMRUGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Binbin Zhou

Physical Scientist

Lynker at NOAA/NWS/NCEP/EMC

5830 University Research Ct.

College Park, MD 20740

***@***.***

301-683-3683

PerryShafran-NOAA commented:
I'm using the directory /lfs/h2/emc/vpppg/noscrub/emc.vpppg/com/evs/v1.0/stats/global_ens.

PerryShafran-NOAA commented:
Oh wait, you can't run on Dogwood....

BinbinZhou-NOAA commented:
I run it in non-mpi mode, just set run_mpi=no in the driver script.

Binbin

On Tue, Jan 23, 2024 at 3:05 PM PerryShafran-NOAA ***@***.***>
wrote:

> Oh wait, you can't run on Dogwood....
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EVS/pull/415#issuecomment-1906834985>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQBMPFKGPPLTS6OK3BT4O53YQAJXFAVCNFSM6AAAAABCCHRNHSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBWHAZTIOJYGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Binbin Zhou

Physical Scientist

Lynker at NOAA/NWS/NCEP/EMC

5830 University Research Ct.

College Park, MD 20740

***@***.***

301-683-3683

PerryShafran-NOAA commented:
I have it already set to run_mpi=no, run_entire_year set to yes.  I'm also running the develop code but with the script from your directory that you had told me to use last week when we came up with the numbers last week.

BinbinZhou-NOAA commented:
Perry,

   I finished the headline run and got exactly the same results as yours,
see the file
/lfs/h2/emc/vpppg/noscrub/binbin.zhou/evs/v1.0/plots/global_ens/headline.20240116/day_acc_below_0.6.txt

In addition, I found the reason why there is some difference between
today's run and last week's run.
The reason is we ran the job on 20240117. This means the the stat data
files on 20240116 were not available
yet. To repeat the run of last week, I set the line 106 in the ush
file evs_global_ens_headline_plot.sh
from days='01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16'
  days='01 02 03 04 05 06 07 08 09 10 11 12 13 14 15'

Then the output results are exactly the same as those in the run of last
week.
But the differences are very small, so the results of both runs should be
acceptable.

Binbin


On Tue, Jan 23, 2024 at 3:19 PM PerryShafran-NOAA ***@***.***>
wrote:

> I have it already set to run_mpi=no, run_entire_year set to yes. I'm also
> running the develop code but with the script from your directory that you
> had told me to use last week when we came up with the numbers last week.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EVS/pull/415#issuecomment-1906854343>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQBMPFNZDPKTN5P5VBAZY3TYQALLRAVCNFSM6AAAAABCCHRNHSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBWHA2TIMZUGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Binbin Zhou

Physical Scientist

Lynker at NOAA/NWS/NCEP/EMC

5830 University Research Ct.

College Park, MD 20740

***@***.***

301-683-3683

PerryShafran-NOAA commented:
Oh, interesting!  Thanks for that info!

PerryShafran-NOAA commented:
@BinbinZhou-NOAA Now that we know that this code works as expected, we are ready to merge it.  Can you do a sync fork for me so we're up to date?

BinbinZhou-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/headline_bugfix

BinbinZhou-NOAA commented:
Perry,

  Yes, I have synced.

Thanks!

Binbin


On Wed, Jan 24, 2024 at 12:54 PM PerryShafran-NOAA ***@***.***>
wrote:

> @BinbinZhou-NOAA <https://github.com/BinbinZhou-NOAA> Now that we know
> that this code works as expected, we are ready to merge it. Can you do a
> sync fork for me so we're up to date?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EVS/pull/415#issuecomment-1908643021>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQBMPFJPT7ROYF3TLXIHE7TYQFDDLAVCNFSM6AAAAABCCHRNHSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBYGY2DGMBSGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Binbin Zhou

Physical Scientist

Lynker at NOAA/NWS/NCEP/EMC

5830 University Research Ct.

College Park, MD 20740

***@***.***

301-683-3683

PerryShafran-NOAA reviewed (approved):
Code works as expected.  Approving.
